### PR TITLE
Recursive CTEs: Stage 1 - add config flag 

### PR DESCRIPTION
**Pull request by matthewgapp**
This PR is the first step in a series of PRs to fully implement recursive CTEs in DF. This PR adds a config flag that will be used to gate further implementation of recursive CTE logical and execution plans

## Which issue does this PR close?

Working toward closing #462

### CODE REVIEW

**Comment by alamb** [comment]
Thanks @matthewgapp  -- what would you think about filing a ticket with the implementation plan described in https://github.com/apache/arrow-datafusion/pull/7581#issuecomment-1886010269 (or maybe we can put it on https://github.com/apache/arrow-datafusion/issues/462 itself)?

**Comment by alamb** [comment]
I think some tests need updating to get the CI run clean here

**Comment by matthewgapp** [comment]
> > I think some tests need updating to get the CI run clean here
> 
> 
> 
> Yeah, this is still a draft, so not quite ready for CI. I'm going to be readying these PRs this afternoon and will mark them Ready for review when done. 

Spent most of my time today digging into the bug fix and repo of the bug(s) we're seeing downstream. Hope to finish that and then clean up these PRs this weekend.

**Comment by matthewgapp** [comment]
> > > I think some tests need updating to get the CI run clean here
> > 
> > 
> > Yeah, this is still a draft, so not quite ready for CI. I'm going to be readying these PRs this afternoon and will mark them Ready for review when done.
> 
> Spent most of my time today digging into the repro and solution to the bug(s) we're seeing downstream. Hope to finish that and then clean up these PRs this weekend.

Thinking that I'll have these three PRs ready for review sometime tomorrow afternoon.

**Comment by matthewgapp** [comment]
> Thinking that I'll have these three PRs ready for review sometime tomorrow afternoon.

Think it's ready for review :)

File: testing
[the remark this answers is not in the document]
Comment (alamb) [reply testing:1]: I think you could potentially update the submodule in main

Mabye like
```shell
cd datafusion
cd testing
# checkout submodule at version that datafusion/main is checked out
git checkout 98fceec
cd ..
git commit -a -m 'restore testing pin'
```

[the remark this answers is not in the document]
Comment (matthewgapp) [reply testing:1]: Thank you! That worked. Think I better understand submodules now after fighting with it for a bit 🙃

**Review by alamb — APPROVED** [review]
Thank you @matthewgapp

**Pull request merged**